From f211d3b2903dbf697ae6a8694c872caae05bf55c Mon Sep 17 00:00:00 2001
From: "eugene.todoruk" <eugene.todoruk@lge.com>
Date: Thu, 28 Mar 2019 17:35:24 +0300
Subject: [PATCH] SMACK: add loading unconfined label

drop FOREACH_LINE() in systemd version 244, it is replaced.

| ../git/src/core/smack-setup.c: In function 'write_unconfined':
| ../git/src/core/smack-setup.c:41:9: error: implicit declaration of function 'FOREACH_LINE'; did you mean 'FOREACH_WORD'? [-Werror=implicit-function-declaration]
|    41 |         FOREACH_LINE(unconfined, f,log_error_errno(errno, "Failed to read line from '%s': %m", "/etc/smack/unconfined")) {
|       |         ^~~~~~~~~~~~
|       |         FOREACH_WORD
| ../git/src/core/smack-setup.c:41:9: warning: nested extern declaration of 'FOREACH_LINE' [-Wnested-externs]
| ../git/src/core/smack-setup.c:41:121: error: expected ';' before '{' token
|    41 |         FOREACH_LINE(unconfined, f,log_error_errno(errno, "Failed to read line from '%s': %m", "/etc/smack/unconfined")) {
|       |
---
 src/core/smack-setup.c | 68 ++++++++++++++++++++++++++++++++++++++++++
 1 file changed, 68 insertions(+)

diff --git a/src/core/smack-setup.c b/src/core/smack-setup.c
index 49b37aefc7..41563b0c45 100644
--- a/src/core/smack-setup.c
+++ b/src/core/smack-setup.c
@@ -25,6 +25,57 @@
 
 #if ENABLE_SMACK
 
+static int write_unconfined(void) {
+        _cleanup_close_ int unconfined_fd = -1;
+        _cleanup_fclose_ FILE *f = NULL;
+        char unconfined[NAME_MAX];
+        size_t len = 0;
+        int r;
+
+        f = fopen("/etc/smack/unconfined", "re");
+        if (!f) {
+                if (errno != ENOENT)
+                        log_warning_errno(errno, "Failed to read '/etc/smack/unconfined': %m");
+
+                return errno == ENOENT ? ENOENT : -errno;
+        }
+
+        for(;;) {
+                _cleanup_free_ char *line = NULL;
+                bool invalid_line = false;
+                int k;
+
+                k = read_line(f, LONG_LINE_MAX, &line);
+                if (k < 0)
+                        return log_error_errno(k, "Failed to read line from '%s': %m", "/etc/smack/unconfined");
+                if (k == 0)
+                        break;
+
+                if (isempty(truncate_nl(unconfined)))
+                        continue;
+
+                len = strlen(unconfined);
+
+                break;
+        }
+
+        if (len == 0)
+                return 0;
+
+        unconfined_fd = open("/sys/fs/smackfs/unconfined", O_WRONLY|O_CLOEXEC|O_NONBLOCK|O_NOCTTY);
+        if (unconfined_fd < 0) {
+                if (errno != ENOENT)
+                        log_warning_errno(errno, "Failed to open '/sys/fs/smackfs/unconfined': %m");
+                return -errno; /* negative error */
+        }
+
+        r = write(unconfined_fd, unconfined, len);
+        if (r < 0)
+                return log_error_errno(errno, "Failed to write unconfined label(%s) to '/sys/fs/smackfs/unconfined': %m", unconfined);
+
+        return 0;
+}
+
 static int write_access2_rules(const char* srcdir) {
         _cleanup_close_ int load2_fd = -1, change_fd = -1;
         _cleanup_closedir_ DIR *dir = NULL;
@@ -413,6 +464,23 @@ int mac_smack_setup(bool *loaded_policy) {
                 return r;
         }
 
+        r = write_unconfined();
+        switch(r) {
+        case -ENOENT:
+                log_debug("Smack is not enabled in the kernel.");
+                break;
+        case ENOENT:
+                log_debug("Smack unconfined file '/etc/smack/unconfined' not found");
+                break;
+        case 0:
+                log_info("Successfully wrote Smack unconfined label.");
+                break;
+        default:
+                log_emergency_errno(r, "Failed to write Smack unconfined label: %m");
+                return r;
+        }
+
+
         *loaded_policy = true;
 
 #endif
